### ISSUE TITLE
Incorrect suggestion for `needless_return_with_question_mark` with `let else`

### ISSUE THREAD

**Issue by Cerber-Ursi** [rust-lang/rust-clippy#11765]
### Summary

In `let-else` construct, `else` branch must always diverge. Sometimes, it is done by returning an error. Sometimes, this error can be converted to the upper-layer error with the question mark. Clippy doesn't like `return` and a question mark in the same line, but it suggests to drop the `return`, which causes compilation error.

Initially found [on URLO](https://users.rust-lang.org/t/analyzer-quirk/102200).

### Lint Name

needless_return_with_question_mark

### Reproducer

I tried this code:

```rust
fn foo() -> Result<(), ()> {
    let Some(s) = Some("") else {
        return Err(())?;
    };
    
    Ok(())
}
```

I saw this happen:

```
warning: unneeded `return` statement with `?` operator
 --> src/lib.rs:3:9
  |
3 |         return Err(())?;
  |         ^^^^^^^ help: remove it
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_return_with_question_mark
  = note: `#[warn(clippy::needless_return_with_question_mark)]` on by default
```

Problem is that the suggestion here is incorrect - it will prevent code from compiling, since `else` branch of `let-else` will no longer diverge.

It _might_ be a good idea to suggest `return Err(()).map_err(Into::into)` instead, but as-is it's a false positive.

### Additional Labels

@rustbot label +I-suggestion-causes-error
